### PR TITLE
Histogram Aggregation - Prometheus exporter

### PR DESCRIPTION
**Pull request by beanliu**
follow up of #2581 
- implement histogram aggregation export in Prometheus.

### CODE REVIEW

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/HistogramAccumulation.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+   * Creates a new {@link HistogramAccumulation} with the given values.
+   *
+   * @return a new {@link HistogramAccumulation} with the given values.
+   */
+  static HistogramAccumulation create(
+      long count, double sum, ImmutableDoubleArray boundaries, ImmutableLongArray counts) {
+    return new AutoValue_HistogramAccumulation(count, sum, boundaries, counts);
+  }
+
+  HistogramAccumulation() {}
```
Comment (aimiebell) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/HistogramAccumulation.java:28]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/DoubleHistogramData.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+package io.opentelemetry.sdk.metrics.data;
+
+import com.google.auto.value.AutoValue;
+import java.util.Collection;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+@AutoValue
+public abstract class DoubleHistogramData implements Data<DoubleHistogramPointData> {
+  DoubleHistogramData() {}
```
Comment (aimiebell) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/DoubleHistogramData.java:15]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/DoubleHistogramPointData.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      Labels labels,
+      double sum,
+      long count,
+      ImmutableDoubleArray boundaries,
+      ImmutableLongArray counts) {
+    return new AutoValue_DoubleHistogramPointData(
+        startEpochNanos, epochNanos, labels, sum, count, boundaries, counts);
+  }
+
+  DoubleHistogramPointData() {}
```
Comment (aimiebell) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/DoubleHistogramPointData.java:45]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/MetricData.java [hunk 4]
```diff
@@ -140,6 +142,28 @@ public static MetricData createDoubleSummary(
...
+        resource,
+        instrumentationLibraryInfo,
+        name,
+        description,
+        unit,
+        MetricDataType.HISTOGRAM,
+        data);
+  }
+
   MetricData() {}
```
Comment (aimiebell) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/data/MetricData.java:167]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,8 @@
 
 ### General
 
+- Support fixed bucket histogram aggregation and exporting them with OTLP/Prometheus exporter.
```
Comment (jkwatson) [inline CHANGELOG.md:7]: this should go down in the metrics section, rather than General

Comment (beanliu) [reply CHANGELOG.md:7]: updated.

**Comment by bogdandrutu** [comment]
Needs a rebase

**Comment by beanliu** [comment]
the PR has been rebased and updated according to the latest changes.

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java [hunk 6]
```diff
@@ -174,6 +178,54 @@ private static void addSummarySamples(
...
+            doubleHistogramPointData.getCount()));
+    samples.add(
+        new Sample(
+            name + SAMPLE_SUFFIX_SUM, labelNames, labelValues, doubleHistogramPointData.getSum()));
+
+    List<String> labelNamesWithLe = new ArrayList<>(labelNames.size() + 1);
+    labelNamesWithLe.addAll(labelNames);
+    labelNamesWithLe.add(LABEL_NAME_LE);
+
+    long[] cumulativeCount = new long[] {0};
```
Comment (bogdandrutu) [inline exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java:201]: Why do you need an array to only use the first value?

**Review by bogdandrutu — APPROVED** [review]
Please address the comment :)

File: exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java (hunk 6 above)
Comment (beanliu) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java:201]: variables used in lambda expression should be final or effectively final. so a primitive here won't do. the alternative would be `AtomicLong`. but not much difference.

Comment (bogdandrutu) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java:201]: :)) can then avoid lambdas :)) I find this very strange. I saw the AtomicLong or the LongAdder

Comment (beanliu) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java]: updated. found a way to rewrite the loop without lambdas.

Comment (bogdandrutu) [reply exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java]: ooo :) nice, there was life before lambdas :))

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**